### ISSUE TITLE
Starting the container fails randomly

### ISSUE THREAD

**Issue by nishantgeorge** [Grokzen/docker-redis-cluster#59]
After running `$ docker run -it -p 7000:7000 grokzen/redis-cluster` to start the container, running `$ redis-cli -c -p 7000 set aaa bbb` Gives the output:
`(error) CLUSTERDOWN Hash slot not served`

So far I haven't found any kind of pattern to explain this. I discovered it because my CI tests started failing today

Here's the output of the `docker run -it ...` command above:
```
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
WARNING: redis-trib.rb is not longer available!
You should use redis-cli instead.

All commands and features belonging to redis-trib.rb have been moved
to redis-cli.
In order to use them you should call redis-cli with the --cluster
option followed by the subcommand name, arguments and options.

Use the following syntax:
redis-cli --cluster SUBCOMMAND [ARGUMENTS] [OPTIONS]

Example:
redis-cli --cluster create 172.17.0.2:7000 172.17.0.2:7001 172.17.0.2:7002 172.17.0.2:7003 172.17.0.2:7004 172.17.0.2:7005 --cluster-replicas 1

To get help about all subcommands, type:
redis-cli --cluster help

==> /var/log/supervisor/redis-1.log <==
43:C 24 Oct 2018 17:07:05.964 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
43:C 24 Oct 2018 17:07:05.964 # Redis version=5.0.0, bits=64, commit=00000000, modified=0, pid=43, just started
43:C 24 Oct 2018 17:07:05.964 # Configuration loaded
43:M 24 Oct 2018 17:07:05.966 * No cluster configuration found, I'm db6288c6719aad403741d564c71875058702f983
43:M 24 Oct 2018 17:07:05.978 * Running mode=cluster, port=7000.
43:M 24 Oct 2018 17:07:05.978 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
43:M 24 Oct 2018 17:07:05.978 # Server initialized
43:M 24 Oct 2018 17:07:05.978 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
43:M 24 Oct 2018 17:07:05.978 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
43:M 24 Oct 2018 17:07:05.978 * Ready to accept connections

==> /var/log/supervisor/redis-2.log <==
45:C 24 Oct 2018 17:07:05.968 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
45:C 24 Oct 2018 17:07:05.968 # Redis version=5.0.0, bits=64, commit=00000000, modified=0, pid=45, just started
45:C 24 Oct 2018 17:07:05.968 # Configuration loaded
45:M 24 Oct 2018 17:07:05.969 * No cluster configuration found, I'm f00b0451cdd3c434cf9e9bc152d06537ddd4ded0
45:M 24 Oct 2018 17:07:05.978 * Running mode=cluster, port=7001.
45:M 24 Oct 2018 17:07:05.978 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
45:M 24 Oct 2018 17:07:05.978 # Server initialized
45:M 24 Oct 2018 17:07:05.978 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
45:M 24 Oct 2018 17:07:05.978 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
45:M 24 Oct 2018 17:07:05.978 * Ready to accept connections

==> /var/log/supervisor/redis-3.log <==
44:C 24 Oct 2018 17:07:05.968 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
44:C 24 Oct 2018 17:07:05.968 # Redis version=5.0.0, bits=64, commit=00000000, modified=0, pid=44, just started
44:C 24 Oct 2018 17:07:05.968 # Configuration loaded
44:M 24 Oct 2018 17:07:05.969 * No cluster configuration found, I'm 1364990a69aabdddf47cc0d49d4ff3f503585fd8
44:M 24 Oct 2018 17:07:05.979 * Running mode=cluster, port=7002.
44:M 24 Oct 2018 17:07:05.979 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
44:M 24 Oct 2018 17:07:05.979 # Server initialized
44:M 24 Oct 2018 17:07:05.979 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
44:M 24 Oct 2018 17:07:05.979 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
44:M 24 Oct 2018 17:07:05.979 * Ready to accept connections

==> /var/log/supervisor/redis-4.log <==
41:C 24 Oct 2018 17:07:05.961 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
41:C 24 Oct 2018 17:07:05.961 # Redis version=5.0.0, bits=64, commit=00000000, modified=0, pid=41, just started
41:C 24 Oct 2018 17:07:05.961 # Configuration loaded
41:M 24 Oct 2018 17:07:05.962 * No cluster configuration found, I'm f11db2ec376ac15af46856f6d451755f954ece43
41:M 24 Oct 2018 17:07:05.970 * Running mode=cluster, port=7003.
41:M 24 Oct 2018 17:07:05.970 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
41:M 24 Oct 2018 17:07:05.970 # Server initialized
41:M 24 Oct 2018 17:07:05.970 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
41:M 24 Oct 2018 17:07:05.970 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
41:M 24 Oct 2018 17:07:05.970 * Ready to accept connections

==> /var/log/supervisor/redis-5.log <==
40:C 24 Oct 2018 17:07:05.961 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
40:C 24 Oct 2018 17:07:05.961 # Redis version=5.0.0, bits=64, commit=00000000, modified=0, pid=40, just started
40:C 24 Oct 2018 17:07:05.961 # Configuration loaded
40:M 24 Oct 2018 17:07:05.962 * No cluster configuration found, I'm 63338abf025fc81859f95d3a78d4c08fbfb66870
40:M 24 Oct 2018 17:07:05.970 * Running mode=cluster, port=7004.
40:M 24 Oct 2018 17:07:05.970 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
40:M 24 Oct 2018 17:07:05.970 # Server initialized
40:M 24 Oct 2018 17:07:05.970 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
40:M 24 Oct 2018 17:07:05.970 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
40:M 24 Oct 2018 17:07:05.970 * Ready to accept connections

==> /var/log/supervisor/redis-6.log <==
42:C 24 Oct 2018 17:07:05.964 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
42:C 24 Oct 2018 17:07:05.964 # Redis version=5.0.0, bits=64, commit=00000000, modified=0, pid=42, just started
42:C 24 Oct 2018 17:07:05.964 # Configuration loaded
42:M 24 Oct 2018 17:07:05.966 * No cluster configuration found, I'm 6594df700e1f2710fcebd8dc57be4cdd79dbeba4
42:M 24 Oct 2018 17:07:05.978 * Running mode=cluster, port=7005.
42:M 24 Oct 2018 17:07:05.978 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
42:M 24 Oct 2018 17:07:05.978 # Server initialized
42:M 24 Oct 2018 17:07:05.978 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
42:M 24 Oct 2018 17:07:05.978 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
42:M 24 Oct 2018 17:07:05.978 * Ready to accept connections
71:X 24 Oct 17:07:08.056 # Fatal error, can't open config file '/redis-conf/sentinel-7000.conf'
72:X 24 Oct 17:07:08.056 # Fatal error, can't open config file '/redis-conf/sentinel-7000.conf'
70:X 24 Oct 17:07:08.056 # Fatal error, can't open config file '/redis-conf/sentinel-7000.conf'
```

**Comment by nishantgeorge** [comment]
Ok, I this issue doesn't seem to exist up till and including 7e9d64ce78643009abace3cf6906db5fd23035fb.
25101d8aee2400c1e73fd40bad7702265472efc2 seems to have introduced the issue.

**Comment by asdptkt** [comment]
Don't know if this is related but when I `make up`, this is the result:

```
127.0.0.1:7000> cluster info
cluster_state:fail
cluster_slots_assigned:0
cluster_slots_ok:0
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:1
cluster_size:0
cluster_current_epoch:0
cluster_my_epoch:0
cluster_stats_messages_sent:0
cluster_stats_messages_received:0
```

**Comment by Grokzen** [comment]
Yeah i see the error. Will try to get a fix out. But if the error is as linked in #60 with redis-trib.rb changes, the old 3.0-4.0 containers should work at least.

**Comment by asdptkt** [comment]
Note that I was able to create the cluster anyway with

```
redis-cli --cluster create 0.0.0.0:7000 0.0.0.0:7001 0.0.0.0:7002 0.0.0.0:7003 0.0.0.0:7004 0.0.0.0:7005 --cluster-replicas 1
```
 
though

**Comment by Grokzen** [comment]
Yes that works as that is the intended way since 5.0 it looks like. Currently redis-trib.rb is used here https://github.com/Grokzen/docker-redis-cluster/blob/master/docker-data/docker-entrypoint.sh#L45 to create the cluster. I need to add some simple if/else case in the entrypoint script that checks either what version of redis-server that is used, or what command is available.

**Comment by asdptkt** [comment]
> @asdptkt I had issues with the redis-cli command on my end to make the cluster create work. I used the same line as you and the output that redis itself recommends to use to create a cluster through redis-cli. But all i get is an error with bad/wrong number of arguments. I will step out from redis 5.0.0 for now and revert back to 4.0.11 as latest and primary redis for this repo. When i get 5.0.0 issues sorted out i will move forward again.

What is your error messages? Maybe I could help.

**Comment by Grokzen** [comment]
@asdptkt Update to this inside the Dockerfile `ARG redis_version=5.0.0` then run `make build up`

Then it outputs the following from the old redis-trib.rb script

```redis-cluster_1  | bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)              
redis-cluster_1  | WARNING: redis-trib.rb is not longer available!            
redis-cluster_1  | You should use redis-cli instead.                                                              
redis-cluster_1  |                                                                                                
redis-cluster_1  | All commands and features belonging to redis-trib.rb have been moved                           
redis-cluster_1  | to redis-cli.
redis-cluster_1  | In order to use them you should call redis-cli with the --cluster
redis-cluster_1  | option followed by the subcommand name, arguments and options.
redis-cluster_1  |
redis-cluster_1  | Use the following syntax:
redis-cluster_1  | redis-cli --cluster SUBCOMMAND [ARGUMENTS] [OPTIONS]
redis-cluster_1  |
redis-cluster_1  | Example:
redis-cluster_1  | redis-cli --cluster create 172.20.0.2:7000 172.20.0.2:7001 172.20.0.2:7002 172.20.0.2:7003 172.20.0.2:7004 172.20.0.2:7005 --cluster-replicas 1
redis-cluster_1  |
redis-cluster_1  | To get help about all subcommands, type:
redis-cluster_1  | redis-cli --cluster help
redis-cluster_1  |
```

So i take the command it suggest me to run and i get the following error

```
root@server:/data# redis-cli --cluster create 172.20.0.2:7000 172.20.0.2:7001 172.20.0.2:7002 172.20.0.2:7003 172.20.0.2:7004 172.20.0.2:7005 --cluster-replicas 1
Unrecognized option or bad number of args for: '--cluster'

root@server:/data# redis-cli --cluster
Unrecognized option or bad number of args for: '--cluster'

root@server:/data# redis-cli --version
redis-cli 3.2.12
```

I used the same command you pasted that "should" work for you, but i get the same error when running that.

**Comment by Grokzen** [comment]
I made an issue upstream to redis to ask if they have any suggestion on how to solve this issue.

**Comment by nishantgeorge** [comment]
@Grokzen Are you sure you're using the version of redis-cli from the 5.0 repo?

**Comment by Grokzen** [comment]
@nishantgeorge @asdptkt  Yes this commit fixes it 23d1f63911c1314fc3158c9c97ccfd1be9b0712c and new images has been pushed upstream. If anyone can verify that the latest and 5.0.0 tag on dockerhub is working as expected, i can close this issue.

**Comment by Grokzen** [comment]
Great. Closing this issue as resolved.
